### ISSUE TITLE
Promote literals in unannotated module globals

### ISSUE THREAD

**Issue by JanFontanet** [astral-sh/ty#1822]
### Summary

I've a 3rd party library that exposes some configuration through module variables, so I've to overwrite them in order for the library to work as I want. But when I run `ty check` on my code I got an error that can be reproduced with the minimal example below.

a.py:
```python
MY_VAR=1
```
main.py
```python
import a

a.MY_VAR=2
```

I got the following error:
```
Object of type `Literal[2]` is not assignable to attribute `MY_VAR` of type `Literal[1]` (invalid-assignment) [Ln 3, Col 1]
```

### Version

ty 0.0.1-alpha.32

**Comment by carljm** [comment]
Thanks for the report!

We probably do need to do literal-promotion on un-annotated module globals, now that we aren't unioning them with `Unknown`.

**Comment by AlexWaygood** [comment]
There was a lot of previous discussion around this in @sharkdp's PRs https://github.com/astral-sh/ruff/pull/20643 and https://github.com/astral-sh/ruff/pull/20664. I support experimenting with this, but I think there are definitely downsides as well as upsides to doing this.

**Comment by sinon** [comment]
To add another datapoint for this. 

The locust load testing framework handles some of it's configuration using this pattern:
https://docs.locust.io/en/stable/configuration.html#customization-of-statistics-settings

Which leads to the same weird error of:
```
error[invalid-assignment] Object of type `Literal[6]` is not assignable to attribute `CONSOLE_STATS_INTERVAL_SEC` of type `Literal[2]`
```

It's easily worked around by re-declaring the value as `Final[int]` / `int` or similar but the error message does look a bit weird and it definetly feels like a rather strict interpretation when the upstream value in locust is untyped.

**Comment by carljm** [comment]
Assuming our principle (the new one! not the old gradual-guarantee one 😆 ) is that we want to assume the more common case and require annotations for the less-common case, it seems like the core question here is which is more common, for an un-annotated module global:
* cases where a precise literal type has significant value over the promoted type (these should be ideally annotated with `: Final`)
* cases where the module global is externally modified (these should be ideally annotated with the full type, accounting for all supported modifications)

My intuition (and, it seems like, @sharkdp's ecosystem experience) suggests that the former might be more common, so it might be the best option to keep the status quo and just require an annotation if mutation is intended.

This is inconsistent with our handling of instance attributes -- but that inconsistency might be consistent with the deeper principle to favor the common case, since instance attributes are almost always set to different values on different instances, whereas module globals are not.

It's worth noting that the precise literal type is more useful to us than to other type checkers, because we support https://docs.astral.sh/ty/features/type-system/#reachability-based-on-types , which other type checkers do not. And @sharkdp did run into real ecosystem cases benefiting from that, where promoting module globals would cause regressions.

There was a suggestion that, in at least one codebase, it was common to have module globals initialized to `None` and later set to other values; we could apply just the widening of `None` to `None | Unknown` to accommodate that. But I'm not convinced; `None | Unknown` isn't a great type, I think I'd prefer to get the nudge to annotate it properly, in that case.

**Comment by carljm** [comment]
I'm going to go ahead and call this: we have ecosystem evidence that doing this would cause more problems than it solves, let's not do it.
